### PR TITLE
Upgrade to TF 2.2

### PR DESCRIPTION
**Pull request by lgeiger**
TensorFlow 2.2.0 will be released 🔜 . This PR upgrades the TensorFlow dependency to point to the 2.2 release branch of TensorFlow. There've been quite a few changes between the two versions so this requires a thorough review.

## TODO
- [x] Get builds to pass
- [x] Upgrade to TensorFlow 2.2.0 stable
- [x] Speed up CI builds since this doubles build time for the converter and End2End tests #287
- [x] Update ImageNet eval tool
- [x] Benchmark models to ensure we don't have performance regressions
- [x] Switch to new Ruy profiler https://github.com/tensorflow/tensorflow/commit/b0f057d1b4ef78f0bba7e8893debe6b4310711ee

## Benchmarks

I ran some preliminary benchmarks on my personal phone for 500 runs with a QuickNet like model and didn't see any performance regressions:

branch | 1 thread (ms) | 4 threads (ms)
--- | --- | ---
master | 26.6 ± 0.3 | 17.5 ± 0.5
PR | 26.5 ± 0.2 | 17.2 ± 0.8

### CODE REVIEW

**Comment by lgeiger** [comment]
TensorFlow made [some changes to `ruy`](https://github.com/tensorflow/tensorflow/commits/master/tensorflow/lite/experimental/ruy) which breaks our kernels. @arashb If you have time it would be cool if you could take a look at it at some point. This is definitely not urgent since TF 2.2 is not released yet.

**Comment by arashb** [comment]
> TensorFlow made [some changes to `ruy`](https://github.com/tensorflow/tensorflow/commits/master/tensorflow/lite/experimental/ruy) which breaks our kernels. @arashb If you have time it would be cool if you could take a look at it at some point. This is definitely not urgent since TF 2.2 is not released yet.

ok, I will take a look. thanks for the headsup 👍

**Comment by AdamHillier** [comment]
The build issues seem to be from this commit: https://github.com/tensorflow/tensorflow/commit/113a37348f295f4858df470f92e4ab49ac0fed23.

I've made https://github.com/larq/compute-engine/pull/286 (a PR into this PR branch) which fixes those errors.

**Comment by lgeiger** [comment]
This PR currently fails compiling XNNPACK when building the benchmark binary for Android on my machine:
```
ERROR: /private/var/tmp/_bazel_lukasgeiger/6f8dfd27496fbafbde3ef0aa730ae6bb/external/XNNPACK/BUILD.bazel:1643:1: C++ compilation of rule '@XNNPACK//:neonfp16arith_ukernels' failed (Exit 1)
external/XNNPACK/src/f16-spmm/gen/32x1-neonfp16arith-unroll2.c:62:26: warning: implicit declaration of function 'vfmaq_f16' is invalid in C99 [-Wimplicit-function-declaration]
        vacc01234567x0 = vfmaq_f16(vacc01234567x0, va01234567x0, vb0);
                         ^
external/XNNPACK/src/f16-spmm/gen/32x1-neonfp16arith-unroll2.c:62:24: error: assigning to 'float16x8_t' (vector of 8 'float16_t' values) from incompatible type 'int'
```
I am not sure if it is due to the fact that I am compiling on macOS or if it will be fixed with the stable release.

**Comment by lgeiger** [comment]
> This PR currently fails compiling XNNPACK when building the benchmark binary for Android on my machine

This is fixed in f8beda0f8968c282d7b5be50b50448bb9d99fd04 and I ran a few sanity benchmarks to verify that this PR doesn't introduce performance regressions.

File: configure.sh [hunk 1]
```diff
@@ -147,7 +150,7 @@ build:android_x86_64 --fat_apk_cpu=x86_64
 # The default android SDK/NDK paths are hardcoded here and the user needs
 # to change the paths according to the local configuration or
 # the "install_android.sh" script
-build --action_env ANDROID_NDK_HOME="/tmp/lce_android/ndk/17.2.4988734"
+build --action_env ANDROID_NDK_HOME="/tmp/lce_android/ndk/18.1.5063045"
```
Comment (jneeven) [inline configure.sh]: @lgeiger  is this change necessary? I locally still have the older version and want to experiment with this a bit

Comment (lgeiger) [reply configure.sh]: Yes, on my system it broke when building the benchmarking binary for ARM64 otherwise https://github.com/larq/compute-engine/pull/258#issuecomment-600616052

**Comment by Maratyszcza** [comment]
The error in compiling XNNPACK is due to old compiler. You need at least Android NDK r18b to build XNNPACK.

**Comment by lgeiger** [comment]
> The error in compiling XNNPACK is due to old compiler. You need at least Android NDK r18b to build XNNPACK.

@Maratyszcza thanks for the help! Indeed we were able to fix it by upgrading the compiler in f8beda0f8968c282d7b5be50b50448bb9d99fd04 :+1:

**Comment by Herdi12** [comment]
Like start bonus

File: .bazelversion [hunk 2]
```diff
@@ -1 +1 @@
-0.29.1
+2.0.0
```
Comment (arashb) [inline .bazelversion:1]: is this upgrade required by TF-2.2?

**Comment by AdamHillier** [comment]
> we are using and also the android SDK version, can you please run the benchmarks on both android and RPi and document the results to make sure no performance regression introduced due to this PR?

I benchmarked QNL on the Raspberry Pi and matched the numbers reported on `docs.larq.dev`.

File: .bazelversion (hunk 2 above)
Comment (lgeiger) [reply .bazelversion:1]: Yes, TensorFlow 2.2 uses bazel 2.0.0

**Comment by lgeiger** [comment]
I just upgraded the PR to 2.2.0 stable, this should be good to go from my side now 🚀

**Comment by arashb** [comment]
as soon as we have the benchmark results this can be merged right?

**Comment by lgeiger** [comment]
> as soon as we have the benchmark results this can be merged right?

@AdamHillier ran benchmarks on the RPI yesterday and didn't see any regression: https://github.com/larq/compute-engine/pull/258#issuecomment-623993349

I ran preliminary benchmarks on my Android phone a while ago and couldn't find any regressions either. However I just reran them with a clean build just to be sure and it looks like there is a slight improvement:

branch | 1 thread (ms) | 2 threads (ms)
--- | --- | ---
master | 36.0 ± 0.3 | 26.4 ± 0.3
PR | 35.6 ± 0.2 | 25.5 ± 0.3
PR (with XNNPack) | 33.7 ± 0.3 | 24.6 ± 0.2

**Comment by arashb** [comment]
> > as soon as we have the benchmark results this can be merged right?
> 
> @AdamHillier ran benchmarks on the RPI yesterday and didn't see any regression: [#258 (comment)](https://github.com/larq/compute-engine/pull/258#issuecomment-623993349)
> 
> I ran preliminary benchmarks on my Android phone a while ago and couldn't find any regressions either. However I just reran them with a clean build just to be sure and it looks like there is a slight improvement:
> 
> branch	1 thread (ms)	2 threads (ms)
> master	36.0 ± 0.3	26.4 ± 0.3
> PR	35.6 ± 0.2	25.5 ± 0.3
> PR (with XNNPack)	33.7 ± 0.3	24.6 ± 0.2

which model is this? Non of the quicknet models in our benchmark table are 36ms for single thread!

**Comment by lgeiger** [comment]
> which model is this? Non of the quicknet models in our benchmark table are 36ms for single thread!

Sorry, this is QuicknetLarge on my personal phone.

**Comment by arashb** [comment]
> > which model is this? Non of the quicknet models in our benchmark table are 36ms for single thread!
> 
> Sorry, this is QuicknetLarge on my personal phone.

ok, thanks later I will run the benchmarks on my pixel phone as well.

File: larq_compute_engine/core/BUILD [hunk 3]
```diff
@@ -47,6 +47,7 @@ cc_library(
     hdrs = ["packbits_utils.h"],
     deps = [
         ":packbits",
+        "@org_tensorflow//tensorflow/lite/experimental/ruy/profiler:instrumentation",
```
Comment (Tombana) [inline larq_compute_engine/core/BUILD:50]: Just a reminder to ourselves: in lce-micro we have removed all this profiling stuff from the core utils (since its not in tflite-micro) and moved it elsewhere. So when we merge that, we shouldn't forget to take it out of these BUILD files again.
